keps/135-startup-policy/README.md

```diff
@@ -0,0 +1,315 @@
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API
+```go
+type StartupPolicyType string
+
+const (
+	// WaitForLeaderReady will create worker statefulset after leader statefulset is ready.
+	WaitForLeaderReady StartupPolicyType = "WaitForLeaderReady"
+
+	// Default will create worker statefulsets immediately while leader statefulset is created.
+	DefaultStartupPolicy StartupPolicyType = "Default"
+)
+
+type LeaderWorkerSetSpec struct {
+	// StartupPolicy determines the startup order of the leader statefulSet and worker statefulSets.
+	// +kubebuilder:default=Default
+	// +kubebuilder:validation:Enum={Default,WaitForLeaderReady}
+	StartupPolicy StartupPolicyType `json:"startupPolicy"`
+}
+```
+
+### Implementation
+We don't need to change the logic when `spec.startupPolicy=Default`. We only need to implement the logic when `spec.startupPolicy=WaitForLeaderReady`.
+
+<!-- #### lsw controller
+lws controller will set worker statefulset's replicas=0 before all leader statefulset pods are ready. -->
+
+#### Pod controller
+Pod controller is responsible for creating worker statefulset. When using `WaitForLeaderReady` strategy, we only need to care about when to create worker statefulset.
+Here is the reconcile logic of creating a new lws with `spec.startupPolicy=WaitForLeaderReady`:
+1. A new leaderworkerset was created.
+2. The leader statefulset was created by lws controller but leader pod is not ready yet.
+3. Pod controller watch the leader pod created/updated event and begin to constructs the worker statefulset configuration. Firstly, it will get the leader statefulset latest status to determine if we should create the worker statefulset pods.
```

Comment:
we could construct the worker sts configuration after checking the leader pod is ready

Reply:
+1, create the workers sts only after the leader is ready